<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
#посчитываем колличество вакансий с помощью функции COUNT()
query_3_1 = f'''SELECT COUNT(*)
                FROM public.vacancies
'''
df = pd.read_sql_query(query_3_1, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

Запрос показал, что в таблице 49197 вакансий.

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
query_3_2 = f'''SELECT COUNT(*)
                FROM public.employers
        '''
df = pd.read_sql_query(query_3_2, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

Запрос показал, что в таблице 23501 работодатель.

3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [6]:
query_3_3 = f'''SELECT count(*)
                FROM public.areas
        '''
df = pd.read_sql_query(query_3_3, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

Запрос показал, что в таблице 1362 региона.

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [7]:
query_3_4 = f'''SELECT count(*)
                FROM public.industries
        '''
df = pd.read_sql_query(query_3_4, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

Запрос показал, что в таблице 294 сфер деятельности компаний.

***

С помощью функции COUNT() мы посмотрели на основную информацию в наших таблицах, посчитав колличество вакансий, а также подсчитали колличество различных областей, работодателей и сферы их деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [8]:
# присоеденив таблицу областей к таблице вакансий, посчитаем колличество вакансий в
# в каждой из областей, отсортируемся и посмотрим на пятерку лидеров
query_4_1 = f'''SELECT a.name area,
                COUNT(a.name) cnt
                FROM public.vacancies v
                LEFT JOIN public.areas a ON v.area_id = a.id
                GROUP BY 1
                ORDER BY 2 DESC
                LIMIT 5
        '''
df = pd.read_sql_query(query_4_1, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

Выполнив запрос мы видим, что в пятерку лидеров вошли следующие города: Москва, Санкт-Петербург, 
Минск, Новосибирск, Алматы.

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [9]:
# с помощью оператора WHERE отфильтруем нашу таблицу, найдем непустые значения
# воспользовавшись оператором IS NOT NULL
query_4_2 = f'''SELECT count(*)
                FROM public.vacancies v
                WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
        '''
df = pd.read_sql_query(query_4_2, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

Мы видим, что в 24073 вакансиях заполнено хотя бы одно из двух полей с зарплатой.

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [10]:
# посмотрим на средние значения c помощью агрегатной функции AVG зарплатной вилки
# и окргулим с помощью функции ROUND 
query_4_3 = f'''SELECT ROUND(AVG(v.salary_from)) min_saalry,
                       ROUND(AVG(v.salary_to)) max_salary           
                FROM public.vacancies v
        '''
df = pd.read_sql_query(query_4_3, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


min_saalry  max_salary
0     71065.0    110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [11]:
# найдем количество вакансий для каждого сочетания
# типа рабочего графика (schedule) и типа трудоустройства (employment)
# сгруппируем данные по сочетаниям и отсорттируем сочетания по убыванию
query_4_4 = f'''SELECT v.schedule schedule,
                       v.employment employment, 
                       COUNT(v.schedule) count      
                FROM public.vacancies v
                GROUP BY 1, 2
                ORDER BY 3 DESC
                LIMIT 5
        '''
df = pd.read_sql_query(query_4_4, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0       Полный день     Полная занятость  35367
1  Удаленная работа     Полная занятость   7802
2     Гибкий график     Полная занятость   1593
3  Удаленная работа  Частичная занятость   1312
4    Сменный график     Полная занятость    940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [12]:
query_4_5 = f'''SELECT v.experience,
                       COUNT(v.experience)          
                FROM public.vacancies v
                WHERE v.experience IS NOT NULL
                GROUP BY 1
                ORDER BY 2 
        '''
df = pd.read_sql_query(query_4_5, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

В данном блоке мы детально проанализировали вакансии из нашей таблицы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [13]:
query_5_1 = f'''SELECT e.name employer_name,
                       COUNT(e.name) count_vacancies
                FROM public.vacancies v
                LEFT JOIN public.employers e ON e.id = v.employer_Id
                GROUP BY 1
                ORDER BY 2 DESC 
                LIMIT 5
        '''
df = pd.read_sql_query(query_5_1, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_name  count_vacancies
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

По результатам запроса мы видим, что на первом месте по колличеству вакансий стоит компания Яндекс, а на пятом - Газпром нефть.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [14]:
query_5_2 = f'''SELECT a.name area_name,
                       COUNT(e.id) count_employers,
                       COUNT(v.id) count_vacancies
                FROM public.areas a
                LEFT JOIN public.employers e ON e.area = a.id
                LEFT JOIN public.vacancies v ON a.id = v.area_id
                GROUP BY 1
                ORDER BY 3 ASC, 2 DESC
                LIMIT 1
        '''
df = pd.read_sql_query(query_5_2, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area_name  count_employers  count_vacancies
0    Россия              410                0

Видим, что больше всего работодателей, у которых нет вакансий, указан регион "Россия".

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [15]:
query_5_3 = f'''SELECT e.name employer_name,
                       count(DISTINCT(a.name)) count_vacancies
                FROM public.vacancies v
                LEFT JOIN public.areas a on a.id = v.area_id
                LEFT JOIN public.employers e on e.id = v.employer_id  
                GROUP BY 1 
                ORDER BY 2 DESC
                LIMIT 5
        '''
df = pd.read_sql_query(query_5_3, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_name  count_vacancies
0                  Яндекс              181
1              Ростелеком              152
2              Спецремонт              116
3  Поляков Денис Иванович               88
4               ООО ЕФИН                71

Исходя из запроса видим, что самое большое колличество регионов, в которые публикует вакансии работодатель, у Яндекса.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [16]:
query_5_4 = f'''SELECT COUNT(*)
                FROM public.employers e
                LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id
                WHERE ei.industry_id IS NULL
        '''
df = pd.read_sql_query(query_5_4, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [17]:
query_5_5 = f'''SELECT e.name employer_name
                FROM public.employers_industries ei
                LEFT JOIN public.employers e ON e.id = ei.employer_id
                LEFT JOIN public.industries i ON i.id = ei.industry_id
                GROUP BY 1
                HAVING COUNT(i.name) = 4
                ORDER BY 1
                LIMIT 3
        '''
df = pd.read_sql_query(query_5_5, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_name
0  101 Интернет
1      21vek.by
2          2ГИС

Видим, что на третьем месте находится компания 2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [18]:
query_5_6 = f'''SELECT COUNT(ei.employer_id)
                FROM public.employers_industries ei
                LEFT JOIN public.industries i ON i.id = ei.industry_id
                WHERE i.name = 'Разработка программного обеспечения'
        '''
df = pd.read_sql_query(query_5_6, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   3553

В таблице 3553 работодателя, сфера которых "Разработка программного обеспечения".

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

In [19]:
# воспользуемся функцией read_html, прочитаем таблицу и положим 
# список городов-миллионников в кортеж.
cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')[1]
cities_tuple = tuple(cities['Город'])

In [20]:
# найдем колличество вакансий компании Яндекс в городах-миллионниках,
# а также воспользовавшись оператором UNION объединиим полученные данные с 
# общим колличеством вакансий в городах-миллионниках
query_5_7 = f'''(SELECT a.name area_name,
                        COUNT(a.id) vacancies_count
                FROM public.vacancies v
                LEFT JOIN public.areas a ON a.id = v.area_id
                LEFT JOIN public.employers e ON e.id = v.employer_id
                WHERE (a.name IN {cities_tuple}) AND (e.name LIKE 'Яндекс')
                GROUP BY 1
                ORDER BY 2)
                UNION
                (SELECT 'Total', count(a.id)
                FROM public.vacancies v
                LEFT JOIN public.areas a ON a.id = v.area_id
                LEFT JOIN public.employers e ON e.id = v.employer_id
                WHERE (a.name in {cities_tuple}) AND (e.name LIKE 'Яндекс')
                GROUP BY 1)
                ORDER BY 2
        '''
df = pd.read_sql_query(query_5_7, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area_name  vacancies_count
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5            Казань               25
6    Ростов-на-Дону               25
7               Уфа               26
8            Самара               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            Total              485

Видим, что в городах-миллионниках у компании Яндекс 485 вакансии.

***

Проанализировав работодателей, мы видим, что самое большое колличество вакансий у компании "Яндекс", а именно 1933, также эта компании шире остальных представлена, учитывая регион вакансии, а если брать города-миллионники, то в них представлено 485 вакансий "Яндекса". Также, видим, что в 8419 вакансий не указана сфера деятельности компании.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

In [21]:
query_6_1 = f'''SELECT COUNT(*)
                FROM public.vacancies
                WHERE (LOWER(name) LIKE '%data%') OR
                (LOWER(name) LIKE '%данн%') 
                '''
df = pd.read_sql_query(query_6_1, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

Видим, что вакансий, имеющих отношении к данным в нашей таблице 1771.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [22]:
query_6_2 = f'''SELECT COUNT(*)
                FROM public.vacancies
                WHERE (((LOWER(name) LIKE '%data scientist%') OR
                (LOWER(name) LIKE '%data science%') OR
                LOWER(name) LIKE '%исследователь данных%') OR
                (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
                (LOWER(name) LIKE '%machine learning%') OR
                (LOWER(name) LIKE '%машинн%обучен%%')) AND
                ((LOWER(name) LIKE '%junior%') OR
                (LOWER(experience) LIKE '%нет опыта%') OR
                (LOWER(employment) LIKE '%cтажировка%')) 
                '''
df = pd.read_sql_query(query_6_2, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

Мы нашли 51 подходящую вакансию.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [23]:
query_6_3 = f'''SELECT COUNT(*)
                FROM public.vacancies
                WHERE (key_skills LIKE '%SQL%' OR 
                LOWER(key_skills) LIKE '%postgres') AND
                ((LOWER(name) LIKE '%data scientist%') OR
                (LOWER(name) LIKE '%data science%') OR
                (LOWER(name) LIKE '%исследователь данных%') OR
                (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
                (LOWER(name) LIKE '%machine learning%') OR
                (LOWER(name) LIKE '%машинн%обучен%%'))
                '''
df = pd.read_sql_query(query_6_3, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

Нашли 201 вакансию, где ключевой навык SQL.

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.


In [24]:
query_6_4 = f'''SELECT COUNT(*)
                FROM public.vacancies
                WHERE (LOWER(key_skills) LIKE '%python%') AND
                ((LOWER(name) LIKE '%data scientist%') OR
                (LOWER(name) LIKE '%data science%') OR
                (LOWER(name) LIKE '%исследователь данных%') OR
                (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
                (LOWER(name) LIKE '%machine learning%') OR
                (LOWER(name) LIKE '%машинн%обучен%%'))
                '''
df = pd.read_sql_query(query_6_4, connection)
df

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [25]:
query_6_5 = '''SELECT ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) +1), 2) "Average skills"
               FROM public.vacancies 
               WHERE (LOWER(name) LIKE '%data%scientist%'
               OR LOWER(name) LIKE '%data%science%' 
               OR LOWER(name) LIKE '%исследователь%данных%' 
               OR LOWER(name) LIKE '%machine learning%' 
               OR LOWER(name) LIKE '%машинн%обучен%' 
               OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%'))  
               AND key_skills IS NOT NULL;            
'''

df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Average skills
0            6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [26]:
query_6_6 = '''SELECT ROUND(AVG(COALESCE(((salary_from+salary_to) / 2), salary_from, salary_to, 0)))
               FROM public.vacancies 
               WHERE (LOWER(name) LIKE '%data%scientist%'
               OR LOWER(name) LIKE '%data%science%' 
               OR LOWER(name) LIKE '%исследователь%данных%' 
               OR LOWER(name) LIKE '%machine learning%' 
               OR LOWER(name) LIKE '%машинн%обучен%' 
               OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%'))
               AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
               AND experience = 'От 3 до 6 лет';            
'''

df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

c:\Users\Антон\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0  243115.0

Мы видим, что вакансий, связанных с данными 1771, но всего 51 вакансия, где требуются специалисты в сфере Data sceince, не учитывая вакансии, в которых ищут специалистов уровня junior. Среди вакансий Data scienceв 201 вакансии ищут сотрудников с ключевым навыком SQL, а в 351 вакансии - python, при том, что в среднем, каждая вакансия имеет 6.41 ключевой навык. Также мы видим, что средняя зарплата для сотрудников имеющих опыт работы от 3 до 6 лет 243115 рублей.

***

# Общий вывод по проекту

Подводя итоги проделанной работы, можно сделать вывод, что, чем больше город, чем больше компания, тем больше сотрудников они ищут, что довольно логично. Также мы выяснили, что среди всех вакансих только 3,5% из сферы, связанной с данными, для человека выбирающего будущую сферу работы может быть полезной такая информация, если сравнить ее с колличеством вакансий в других сферах. Также, мы выяснили, что не для всех, и, даже не для половины вакансий, в качестве ключевого навыка требуются python, SQL. Выяснение, какие именно навыки требуются работодателям, также может является полезным, поскольку обучающиеся могут целенаправленно учить именно то, что требуется.